bridges/modules/messages/src/lib.rs

```diff
@@ -271,92 +233,89 @@ pub mod pallet {
-				);
+			let mut lane = active_inbound_lane::<T, I>(lane_id)?;
 
-				if let Some(lane_state) = lane_data.lane_state {
-					let updated_latest_confirmed_nonce = lane.receive_state_update(lane_state);
-					if let Some(updated_latest_confirmed_nonce) = updated_latest_confirmed_nonce {
-						log::trace!(
-							target: LOG_TARGET,
-							"Received lane {:?} state update: latest_confirmed_nonce={}. Unrewarded relayers: {:?}",
-							lane_id,
-							updated_latest_confirmed_nonce,
-							UnrewardedRelayersState::from(&lane.storage_mut().get_or_init_data()),
-						);
-					}
-				}
+			// subtract extra storage proof bytes from the actual PoV size - there may be
+			// less unrewarded relayers than the maximal configured value
+			let lane_extra_proof_size_bytes = lane.storage().extra_proof_size_bytes();
+			actual_weight = actual_weight.set_proof_size(
+				actual_weight.proof_size().saturating_sub(lane_extra_proof_size_bytes),
+			);
 
-				let mut lane_messages_received_status =
-					ReceivedMessages::new(lane_id, Vec::with_capacity(lane_data.messages.len()));
-				for mut message in lane_data.messages {
-					debug_assert_eq!(message.key.lane_id, lane_id);
-					total_messages += 1;
-
-					// ensure that relayer has declared enough weight for dispatching next message
-					// on this lane. We can't dispatch lane messages out-of-order, so if declared
-					// weight is not enough, let's move to next lane
-					let message_dispatch_weight = T::MessageDispatch::dispatch_weight(&mut message);
-					if message_dispatch_weight.any_gt(dispatch_weight_left) {
-						log::trace!(
-							target: LOG_TARGET,
-							"Cannot dispatch any more messages on lane {:?}. Weight: declared={}, left={}",
-							lane_id,
-							message_dispatch_weight,
-							dispatch_weight_left,
-						);
-
-						fail!(Error::<T, I>::InsufficientDispatchWeight);
-					}
+			if let Some(lane_state) = lane_data.lane_state {
+				let updated_latest_confirmed_nonce = lane.receive_state_update(lane_state);
+				if let Some(updated_latest_confirmed_nonce) = updated_latest_confirmed_nonce {
+					log::trace!(
+						target: LOG_TARGET,
+						"Received lane {:?} state update: latest_confirmed_nonce={}. Unrewarded relayers: {:?}",
+						lane_id,
+						updated_latest_confirmed_nonce,
+						UnrewardedRelayersState::from(&lane.storage().data()),
+					);
+				}
+			}
 
-					let receival_result = lane.receive_message::<T::MessageDispatch>(
-						&relayer_id_at_bridged_chain,
-						message.key.nonce,
-						message.data,
+			let mut messages_received_status =
+				ReceivedMessages::new(lane_id, Vec::with_capacity(lane_data.messages.len()));
+			for mut message in lane_data.messages {
+				debug_assert_eq!(message.key.lane_id, lane_id);
+				total_messages += 1;
+
+				// ensure that relayer has declared enough weight for dispatching next message
+				// on this lane. We can't dispatch lane messages out-of-order, so if declared
```

Comment:
I thought messages could be delivered out-of-order. This means if there's a really big message it'll clog the lane until some relayer is willing to send it?

Reply:
I don't remember exact reasoning, but the order is crucial for one lane, e.g. there is a description:
> Messages that are sent over the same lane are guaranteed to be delivered to the target chain in
the same order they're sent from the source chain. In other words, message with nonce `N` will be delivered right before
delivering a message with nonce `N+1`.

I think it is about some guarantees, I think that HRMP/UMP/DMP works also  in the order.
I think when you process message between two points in random order it can allow attacks, e.g. if you send two messages (in order: msg1 and then msg2):
```
msg1: AHP -> deposit(a1) -> AHK
msg2: AHP -> do_whatever_with(a1) -> AHK
```
we cannot just randomly change the order.